Updated prob-interface.md

YongchaoHuang opened the pull request:
1. fixed the invalid `logprior` and `logjoint` functions;
2. added `σ=1` to the demo model as it's described in the context but missing in the original version;
3. fixed `cross_val` accordingly.

NB: as the `logprior` and `logjoint` methods are still in development. here we temporarily built them without strict signatures.

YongchaoHuang commented:
> Can you explain the motivation for the PR? It's a bit unclear to me since the current code works just fine. In contrast to the existing implementation, the new code is also not tested (i.e., not run when the docs are built). You also removed part of the code for dictionaries it seems (which, as stated in the docs, you really want to use instead of NamedTuples if the model has many variables)?

@devmotion Many thanks for questioning. Yes I made the changes based on following considerations:
1. Obvious σ missing in the original model specification. I added σ=1 to the demo model as it's described in the context but missing in the original version;
2. fixed the invalid `logprior` and `logjoint` functions. These functions are used for accepting samples of Dictionary and NamedTuple types (e.g. `logjoint(model, sample)` and `logjoint(model, sample_dict)`), however, I guess they are not working in the original version? With current version of DynamicPPL.jl on my machine, it doesn't have any matching method for `logjoint(model, NamedTuple)`  (at for the specific NamedTuple `sample_nt=rand(model)`), so to make it work two newly built temporary functions with signature `logprior(model, nt)` and `logjoint(model,nt)` were added.
These two functions also work with a vector of NamedTuples, rather than a single NamedTuple.
Could you point me to the source code `logprior(model, Dict)` and `logjoint(model, Dict)`? They are not working on my side as well, so I removed them.

The modified codes are working in general if you run the codes through, but the build fails due to errors induced by `make.jl` - this is something I need to work with you and other team members to understand the design of the tests and why they fail.

devmotion commented:
> Obvious σ missing in the original model specification. I added σ=1 to the demo model as it's described in the context but missing in the original version;

Ah, I see, you mean because we condition on sigma? That's just wrong there, we should remove the conditioning. I think that's simpler than adding a hardcoded 1 to the model.

> * fixed the invalid `logprior` and `logjoint` functions. These functions are used for accepting samples of Dictionary and NamedTuple types (e.g. `logjoint(model, sample)` and `logjoint(model, sample_dict)`), however, I guess they are not working in the original version? With current version of DynamicPPL.jl on my machine, it doesn't have any matching method for `logjoint(model, NamedTuple)`  (at least for the specific NamedTuple `sample_nt=rand(model)`), so to make it work two newly built temporary functions with signature `logprior(model, nt)` and `logjoint(model,nt)` were added.
>   These two functions also work with a vector of NamedTuples, rather than a single NamedTuple.
>   Could you point me to the source code `logprior(model, Dict)` and `logjoint(model, Dict)`? They are not working on my side as well, so I removed them.

I'm suspecting that there might be a problem in your local setup. The documentation, including this tutorial, was built successfully five days ago (and possibly even more recently in a PR, I did not check). The functions are defined here: https://github.com/TuringLang/DynamicPPL.jl/blob/6d5dc4028db6831f6ddbec8544e825a91aad816c/src/simple_varinfo.jl#L553-L647 NamedTuples and dictionaries are packed in a `SimpleVarInfo` which is then used internally for evaluating the desired log density. Also the doctests in the docstrings for these functions passed succesfully only some days ago.

YongchaoHuang commented:
> > Obvious σ missing in the original model specification. I added σ=1 to the demo model as it's described in the context but missing in the original version;
> 
> Ah, I see, you mean because we condition on sigma? That's just wrong there, we should remove the conditioning. I think that's simpler than adding a hardcoded 1 to the model.
> 
> > * fixed the invalid `logprior` and `logjoint` functions. These functions are used for accepting samples of Dictionary and NamedTuple types (e.g. `logjoint(model, sample)` and `logjoint(model, sample_dict)`), however, I guess they are not working in the original version? With current version of DynamicPPL.jl on my machine, it doesn't have any matching method for `logjoint(model, NamedTuple)`  (at least for the specific NamedTuple `sample_nt=rand(model)`), so to make it work two newly built temporary functions with signature `logprior(model, nt)` and `logjoint(model,nt)` were added.
> >   These two functions also work with a vector of NamedTuples, rather than a single NamedTuple.
> >   Could you point me to the source code `logprior(model, Dict)` and `logjoint(model, Dict)`? They are not working on my side as well, so I removed them.
> 
> I'm suspecting that there might be a problem in your local setup. The documentation, including this tutorial, was built successfully five days ago (and possibly even more recently in a PR, I did not check). The functions are defined here:
> 
> https://github.com/TuringLang/DynamicPPL.jl/blob/6d5dc4028db6831f6ddbec8544e825a91aad816c/src/simple_varinfo.jl#L553-L647
> NamedTuples and dictionaries are packed in a `SimpleVarInfo` which is then used internally for evaluating the desired log density. Also the doctests in the docstrings for these functions passed succesfully only some days ago.

@devmotion Thanks very much for the information. Yes, there is a problem on my machine measuring the size of a `NamedTuple`, as shown below: 
```
julia> @model function demo(x)
                  m ~ Normal()
                  for i in eachindex(x)
                      x[i] ~ Normal(m, 1.0)
                  end
              end
demo (generic function with 2 methods)

julia> logprior(demo([1.0]), (m = 100.0, ))
ERROR: MethodError: no method matching size(::NamedTuple{(:m,), Tuple{Float64}}, ::Int64)
Closest candidates are:
  size(::Union{QR, LinearAlgebra.QRCompactWY, QRPivoted}, ::Integer) at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\qr.jl:557
  size(::Union{Cholesky, CholeskyPivoted}, ::Integer) at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\cholesky.jl:443
  size(::Union{Hermitian{T, S}, Symmetric{T, S}} where {T, S}, ::Any) at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.6\LinearAlgebra\src\symmetric.jl:201
  ...
Stacktrace:
 [1] logprior(model_instance::Model{typeof(demo), (:x,), (), (), Tuple{Vector{Float64}}, Tuple{}, DefaultContext}, nt_arr::NamedTuple{(:m,), Tuple{Float64}})
   @ Main c:\Users\h\Desktop\new Julia\src\404_20230120.jl:49
 [2] top-level scope
   @ REPL[277]:1
```
If the signatures `logprior(model, NamedTuple)` and `logjoint(model, Dict)` work, then no need to put the new functions in the loop. 
Shall I just close this PR, and remove the condition on σ in the original version?

yebai commented:
We likely don't need this PR.
